### DISCUSSION TITLE
Project X 需要更多的英雄！

### DISCUSSION THREAD

**Discussion by badO1a5A90** [XTLS/Xray-core#56] [Ideas]
我们需要更多的英雄‼

Project X 还有很多事情可以做，在这里可以查看还需要完善的事情，包括文档部分也需要更多英雄一起来完成。
大家可以看看能做点什么帮助 Project X 变强，比如写点心得PR到文档网址。

非常感谢！🙏

## 性能提升
- 内存占用
  - ~~arm64上启动占用~~(https://github.com/XTLS/Xray-core/issues/68)
  - kcp内存占用多，释放不及时，历史问题
  - 缓存重复利用率提高，降低整体内存占用
- ~~减少rtt,TLS1.3和ws应该都可以减-1.~~ https://github.com/XTLS/Xray-core/pull/375 https://github.com/v2fly/v2ray-core/issues/378
- ds的性能,历史问题 https://github.com/v2fly/v2ray-core/issues/373
- 统计对性能的影响问题,历史问题. https://github.com/XTLS/Xray-core/pull/221 https://github.com/v2fly/v2ray-core/issues/416
- ~~更好的利用多核~~
  - ~~测试下指定超过实际数量的CPU核心数是否能提高效率~~ 目前测试结论为不能 https://github.com/XTLS/Xray-core/discussions/56#discussioncomment-221285
  - ~~解决一核有难多核围观的历史问题~~ https://github.com/XTLS/Xray-core/issues/81

### 标准
- 分享标准 https://github.com/XTLS/Xray-core/issues/91
- UUID https://github.com/XTLS/Xray-core/issues/158
### ~~UDP~~
- ~~解决UDP历史问题~~
  - ~~修改ss trojan的UDP~~ https://github.com/XTLS/Xray-core/commit/8f8f7dd66f1c116d55feaafde38f4c008592a70a
  - ~~修改和增强 ss多倍发包~~
  - ~~VLESS聚合隧道 的UDP~~ 
  - ~~SS trojan UDP fullcone~~ https://github.com/XTLS/Xray-core/releases/tag/v1.2.0
### 核心
- TLS配置增强 https://github.com/v2fly/v2ray-core/issues/375 https://github.com/XTLS/Xray-core/pull/78
  - TLS 指纹
    -[uTLS fingerprints _尚不完善_](https://github.com/XTLS/Xray-core/pull/451)
    -[WSS Browser Dialer](https://github.com/XTLS/Xray-core/pull/421)  
  - ~~version~~
  - ~~Ciphers~~
  - ~~Prefer_server_ciphers~~
  - ~~Curves~~
  - ~~OCSP Stapling 有bug待修复~~
  - ~~自动申请证书，自动加载新证书而不用重启xray~~(有待更多讨论确定方案)
  - ~~CertificateObject / password of key or keyFile~~
  - 炸Ciphers错误配置 @rprx
  - ~~自动热重载证书~~ @eMeab 
  - ~~disable_session_resumption（补充文档）~~
- mux的改进或重新实现或结合smux，解决历史问题。
- ~~SNI分流~~
- DNS
  - ~~Edns-Client-Subnet~~
  - ~~fake DNS~~
 - ~~增加更多的配置格式(TOML/YAML)~~ https://github.com/XTLS/Xray-core/issues/75 https://github.com/XTLS/Xray-core/pull/98 https://github.com/XTLS/Xray-core/pull/86
- 热加载配置
- 协议增强
  - 支持 mtproto 2.0
  - ~~给SS增加多用户~~ 和 api
  - socks inbound 可以接受 http 协议并转发至另一个http inbound
- 传输方式增强
  - [gRPC Transport](https://github.com/XTLS/Xray-core/pull/356)
- 日志增强
  - ~~日志优化~~ https://github.com/XTLS/Xray-core/commit/be9421fedf9017f1ea48c292a2ba7c9d8e3ed9dc
  - 结构化日志
### 负载均衡
- 负载均衡 
  - 择优选择 https://github.com/XTLS/Xray-core/pull/168 https://github.com/v2fly/v2ray-core/issues/377 https://github.com/v2ray/v2ray-core/pull/1924
  - 故障转移 
- 类似clash模式的路由分流优化
  - 分流差一个根据进程分流,
  - 代理组(负载均衡/故障转移),
  - 热切换 
### 其他
- 启动时应该对配置更细致的检查,并输出错误配置,比如trojan协议时,有相同邮箱等.
- ~~修改panic的退出代码,以匹配service的重启方针~~ https://github.com/XTLS/Xray-install/commit/1445230357ec22af4facb6141056cd8538e11a31

## 脚本加强
- ~~改root,建议自定义用户和组~~
- ~~加参数可以不覆盖service文件(避免覆盖用户自己修改)~~
- ~~[更多新功能的版本测试中] (https://github.com/XTLS/Xray-install/tree/dev)~~ 已上线
- ~~脚本中控制service文件,不需要在release中包含~~

### 规范
- 文档编写遵循[中文文案排版指北](https://github.com/sparanoid/chinese-copywriting-guidelines/blob/master/README.zh-CN.md)进行梳理
- ~~规范文档中使用的图片的放置目录~~
- ~~规范文档中 notice 的使用~~
- ~~文档添加许可证（eg. CC BY-NC-SA 4.0~~
### 内容
  - ~~关于~~
  - ~~最近更新~~
  - ~~快速入门~~
  - ~~配置文件~~ 
  - Xray 核心三连 -- XTLS 深度剖析, VLESS协议详解, Fallfack奇技淫巧
  - ~~使用心得~~ 
  - 使用心得陆续增加
    - 各种基本组合技搭配指南
    -~~geo分流实际应用（从此再无PAC~~ 需要增加level2说明嗅探+domainStrategy+useIP等组合用法
    - 证书相关文章 
    - sniffing释疑
    - ~~透明代理配置~~
    - 内置DNS详解与黑科技
    - proxy protocol 详解
    - ~~各种回落配置应用示范~~ 需增加SNI部分
    - API 高级使用手册 https://github.com/XTLS/XTLS.github.io/pull/67
    - Xray使用splice高性能中转杂技
  - ~~QA~~
  -  QA陆续增加
  - 大案牍术
  - links
  - 需要召集更多英雄一起修缮文档！
### 语言翻译
  - 需补全
### 视觉
  - 已经切换至 [next](https://xtls.github.io/Xray-docs-next/)
  - ~~视觉细节调整~~
    -  ~~代码段配色~~
    - ~~各篇文档的细化描述，纠错，排版调优~~
    - ~~更多配色设计方案（求更多设计大佬~~
  - ~~移动终端适配优化~~
  - ~~被墙域名优化~~

## 数据文件
- geosites
  - https://github.com/XTLS/Xray-core/discussions/56#discussioncomment-289794 
- geoIP

## 模板仓库
- ~~透明代理模板~~ 参见[透明代理文档](https://xtls.github.io/documents/level-2/)中的配置实例即可.
- XTLS 最新流控模板( splice )
- TCP, UDP 分开使用的模板
- Xray-tproxy 放置各种场景的iptables及配置规则的脚本

## 发布通道
- 建立尝鲜版本和稳定版本两个发布通道
  - 临时版本,主要用于特定情况的测试(比如从分支build的), 于群内/issue回复等渠道 发布特定版本 
  - 尝鲜版本可以为自动 daily build , 用于尝鲜和获得即时反馈和再改进
  - 稳定版本为定时更新(比如月更), 合并稳定的修改并发布
- 需要自动构建流程

## 开发分支
- 建议新功能开新分支
  - 完成以后合并入主干
  - 完成以后扬了
- cherry-pick

## 性能及功能测试
- 更多的实际环境测试数据收集整理

## 把这个搬到project boards

**Comment by badO1a5A90** [comment]
arm64上启动占用,结案
https://github.com/XTLS/Xray-core/issues/68

**Comment by ghost** [comment]
Waiting for fallback with speed test :)
https://github.com/v2fly/v2ray-core/issues/377#issuecomment-745758286

**Reply by badO1a5A90** [reply]
> Waiting for fallback with speed test :)
> [v2fly/v2ray-core#377 (comment)](https://github.com/v2fly/v2ray-core/issues/377#issuecomment-745758286)

看起来像是基于 https://github.com/v2ray/v2ray-core/pull/1924 的改进?

**Reply by ghost** [reply]
Yes. Looking forward to the integration or a rewrite in Xray

**Comment by badO1a5A90** [comment]
感谢 @emeab 大佬的pr https://github.com/XTLS/Xray-core/pull/78

**可以自定义配置TLS的最低最高版本,以及指定加密套件及顺序偏好**

<br /> 

**TIP:**
**如果要在 ssllibs 或者 myssl 获得 A 等级的评价,你现在可以这样做:**

1. 前提, 你的证书是ECC证书, 比如你用acme.sh可以这样获取
```
./acme.sh --issue -d "your.domain" -w /var/www/html -k ec-256 --cert-file /etc/xray/"your.domain"/"your.domain".cer --key-file /etc/xray/"your.domain"/"your.domain".key --fullchain-file /etc/xray/"your.domain"/fullchain.cer --force

```

<br /> 

2. 然后只需简单的在 **服务端的** TLS 配置 (streamsettings中的tlsSettings/xtlsSettings) 中加入
```
"minVersion": "1.2"
```
**即可在myssl或ssllabs上获得A**

 <br /> 

3. 然后在回落的网站服务器上开启HSTS,如 Nginx 只需配置   
```
add_header Strict-Transport-Security "max-age=63072000" always;
```
**即可在myssl或ssllabs上获得A+**
  
 <br />  <br />  <br />  <br />  <br />  <br />  <br />  <br /> 
------
如果你熟悉加密套件的配置, 你现在可以通过配置指定加密套件顺序及偏好设定
**不熟悉这部分则不需要关心**
**不熟悉这部分则不需要关心**
**不熟悉这部分则不需要关心**



在Xray的 TLS 的配置(streamsettings中的tlsSettings/xtlsSettings) 中加入:
```
"cipherSuites": "此处填写你需要的加密套件名称,每个套件名称之间用:进行分隔"
```
```
"preferServerCipherSuites": true
```
**以上配置为非必要选项，正常情况下不影响安全性**
**在未配置的情况下golang根据设备自动选择**
**因填写不当引起的问题自行负责**
<br />

请注意, 加密套件的名称和nginx等一些软件或有些不同.

你可以在[这里](https://golang.org/src/crypto/tls/cipher_suites.go#L500)或[这里](https://golang.org/src/crypto/tls/cipher_suites.go#L44)找到golang加密套件的名词和说明

**Comment by ghost** [comment]
SNI分流是否有必要做.

**Reply by badO1a5A90** [reply]
> SNI分流是否有必要做.

双手支持

**Comment by badO1a5A90** [comment]
关于CPU多核.

1. 正常build
2. 使用https://github.com/uber-go/automaxprocs
3. 指定超过实际的CPU数量 (有文章阐述在某些情形下可能可以提高性能)
```
 	cpuNum := runtime.NumCPU()
	runtime.GOMAXPROCS(cpuNum * 2)
```

测试环境中 以上测试结果表现均一致

**Comment by vangork** [comment]
开个脑洞，client端集成tun2socks, client和server的连接支持多种加密（GSM，证书），支持icmp，展望一下此project可以PK openvpn，走向成为商业化的vpn解决方案。

**Comment by cncherisher** [comment]
希望可以支持faketls模式的mtproto,这样又可以少装一个软件🤔

**Comment by JimLee1996** [comment]
日志加强，print入站tag和出站tag的信息，即原本的[outbound]改为类似于[inbound -> outbound]。例如下图假设入站名为http_in，[direct]显示成[http_in -> direct]

![image](https://user-images.githubusercontent.com/15334330/103432403-2d219100-4c19-11eb-9e21-5b74603d9482.png)

**Reply by badO1a5A90** [reply]
> 日志加强，print入站tag和出站tag的信息，即原本的[outbound]改为类似于[inbound -> outbound]。例如下图假设入站名为http_in，[direct]显示成[http_in -> direct]
> 
> ![image](https://user-images.githubusercontent.com/15334330/103432403-2d219100-4c19-11eb-9e21-5b74603d9482.png)

我觉得这个建议不错

**Reply by nJhEqnvK** [reply]
非常赞同，特别是这几天排查DNS的时候。log只有一个`1.1.1.1:32[dns_out]`，没法直观看到路由方式就很难受

**Reply by badO1a5A90** [reply]
> 日志加强，print入站tag和出站tag的信息，即原本的[outbound]改为类似于[inbound -> outbound]。例如下图假设入站名为http_in，[direct]显示成[http_in -> direct]
> 
> ![image](https://user-images.githubusercontent.com/15334330/103432403-2d219100-4c19-11eb-9e21-5b74603d9482.png)

https://github.com/XTLS/Xray-core/commit/be9421fedf9017f1ea48c292a2ba7c9d8e3ed9dc

**Comment by BioniCosmos** [comment]
内置 DNS 详解和黑科技在[这篇文章](https://medium.com/@TachyonDevel/%E6%BC%AB%E8%B0%88%E5%90%84%E7%A7%8D%E9%BB%91%E7%A7%91%E6%8A%80%E5%BC%8F-dns-%E6%8A%80%E6%9C%AF%E5%9C%A8%E4%BB%A3%E7%90%86%E7%8E%AF%E5%A2%83%E4%B8%AD%E7%9A%84%E5%BA%94%E7%94%A8-62c50e58cbd0)中就很全了，请问还需要有什么补充？

**Reply by badO1a5A90** [reply]
> 内置 DNS 详解和黑科技在[这篇文章](https://medium.com/@TachyonDevel/%E6%BC%AB%E8%B0%88%E5%90%84%E7%A7%8D%E9%BB%91%E7%A7%91%E6%8A%80%E5%BC%8F-dns-%E6%8A%80%E6%9C%AF%E5%9C%A8%E4%BB%A3%E7%90%86%E7%8E%AF%E5%A2%83%E4%B8%AD%E7%9A%84%E5%BA%94%E7%94%A8-62c50e58cbd0)中就很全了，请问还需要有什么补充？

这篇文章写得非常好且非常详尽，但不知道如何搬运比较好。

**Comment by BioniCosmos** [comment]
我的想法是，作为文档，也许无需太复杂。如果非要添加的话，可以在使用技巧那里写一条外链指向文章，或者添加进来并标明转载。还有一种方式就是再写一篇……

**Comment by badO1a5A90** [comment]
感觉写几组复杂度不同用途不同的配置。
然后外链这篇详解比较好吧。

**Comment by ZZYSonny** [comment]
关于文档有个建议，就是采用json-schema。可以在编辑器中看到有什么是必填的，某个值的文档，某个值可以写哪些东西。

个人认为\*ray配置文件结构复杂，从零写出一个配置文件不仅要看反复看整个文档，还需要有OOP基础（文档都是****Object的）。而json-schema可能会让这个过程更加顺滑一些。

实现方法可以参考https://github.com/cwxyz007/v2ray-config-json-schema
用typescript生成json-schema，比较符合现在配置文件OOP的特点。这样网页可以专注于介绍xray的特色功能。同时考虑是否能同时生成网页版的文档。
但主要问题是，一些xray没文档的用v2fly的文档是否会有版权问题。

//似乎也有从go生成json-schema的工具，但可能用ts的可以避免写文档的工作全都压在开发者身上。

**Comment by zhangsh-cn** [comment]
除了 xray-core 之外，再创建另外两个项目：xray-ui 和 xray-controller

* xray-ui 就是一套跨平台的通用用户界面，可以优先考虑用 HTML/Flutter 来实现。

* xray-controller 主要做一些用户需要而又不太合适在xray-core里实现的功能，包括但不限于：

    1. 给 xray-ui 使用的 Rest API
    2. TUN 的支持
    3. 规则匹配模式
    4. 订阅支持

**Reply by badO1a5A90** [reply]
>   1. 给 xray-ui 使用的 Rest API
>   2. TUN 的支持
>   3. 规则匹配模式
>   4. 订阅支持

~~xray-Flutter 的文件夹已经建好了~~

1 指什么,没明白...
2.很快会有
3.其实就是路由分流了.
4.可以考虑

**Comment by zhangsh-cn** [comment]
老哥，我这四条是指 xray-controller 提供的功能。

**Comment by ricuhkaen** [comment]
**建议增补并细分 geosite.dat 的类型** ：ads-all

适用 block 的类型应该进一步细分，不能只有一个 ads-all，这个过度耦合，较难灵活管理。

可以参考adguard或者ublock origin的分法：

1. **Ads:** _ad networks, per-country sub-list, etc_
2. **Security:** _malware, spamming, coin mining, etc_
3. **Privacy:** _user tracking, fingerprinting, etc_)
4. **Social Media Widgets:** _facebook, twitter, github, instagram, etc_
5. **Annoyances**
6. **Others**

这样可以在不增加任何xray-core代码的情况下，为用户提供类似或者持平 Pi-hole 定制过滤和隐私保护。

这些上游都有开源的清单，但是尚不确定协议问题的影响，待研究。若无问题则可基于上游做转化。

**Comment by ghost** [comment]
目前，`Outbound.ProxySettings`允许用户指定一个tag，将此`Outbound`发出的数据，转发至所指定的`Outbound`发出。但是转发会使`streamSettings` 失效。
于是，有群友提出建议：**使`ProxySettings`支持`streamSettings`**，以提供更加高效与灵活的链式转发。

**Reply by badO1a5A90** [reply]
> 目前，`Outbound.ProxySettings`允许用户指定一个tag，将此`Outbound`发出的数据，转发至所指定的`Outbound`发出。但是转发会使`streamSettings` 失效。
> 于是，有群友提出建议：**使`ProxySettings`支持`streamSettings`**，以提供更加高效与灵活的链式转发。

这个改动有什么实际应用场景吗?

**Reply by ghost** [reply]
链式代理，可用于**不受控制**的中转机。
如某些服务提供商的节点接入了优质线路，可以使用其作为中转。但是被中转的流量目前不使用`streamSettings`，使其不得不采用`vMess`或`shadowsocks`等加密协议，导致性能损失，且无法受益于`xtls`的性能提升。

**Reply by doudouhn6** [reply]
没有链式代理功能的xray仅靠xtls是没有太“神话”般的竞争力的。试想一个架构，在一台拥有iplc路由的主机上安装哪怕是已经支持链式代理的v2ray，那么网速和使用便捷性也远超xray。另一方面，如果主机本身的路由线路就差，那么无论安装什么ray都显得无能为力

**Comment by JimLee1996** [comment]
### 【Feature Request】内置DNS是否可以支持DNS over TCP的查询

今天在使用内置DNS时发现，内置DNS服务器的查询流量（使用最朴素的`8.8.4.4:53`）无法通过`"network"`默认设置为`"tcp"`的shadowsocks隧道传输，会出现context deadline exceeded问题；而设置为`"tcp,udp"`则无问题。（vless隧道中无相关问题。）

请求一个类似于pdnsd的功能，可以选择udp或tcp查询方式的设定配置。如果实现起来比较麻烦就算了。。。

**Reply by RPRX** [reply]
DoH 即可

**Reply by JimLee1996** [reply]
> DoH 即可

DoH的确是走在tcp上的。但是对于远端DNS服务器不支持DoH，内置DNS服务器只能用朴素的方式查询了，而改种方式只支持udp。如隧道不支持udp over tcp，就会出现问题。可以考虑增加一下这样的配置，比如将"8.8.4.4"写成"tcp://8.8.4.4"就开启了tcp查询模式。

**Comment by futureproperty** [comment]
自带tun还是很需要的，有些路由器阉割tproxy
还有
纯UDP数据流也要，udp on tcp导致小包数量直接double，转发吃力，玩个CSGO从来没见到SendQ归0，一直有300-500的包停在SendQ，tcp确认机制更是导致实际延迟上升
